azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -151,211 +127,55 @@ def __get_agent_family_manifests(self, goal_state):
-            agent_family = self.__get_agent_family_manifests(goal_state)
-            requested_version = self.__get_requested_version(agent_family)
-            agent_manifest = None  # This is to make sure fetch agent manifest once per update
-            warn_msg = ""
-            if requested_version is None:
-                # Do not proceed with update if self-update needs to download the manifest again with in an hour
-                if not self.__should_agent_attempt_manifest_download():
-                    return
-                if conf.get_enable_ga_versioning():  # log the warning only when ga versioning is enabled
-                    # TODO: Need to revisit this msg when version is missing in Goal state. We may need to handle better way to report the error
-                    warn_msg = "Missing requested version in agent family: {0} for incarnation: {1}, fallback to largest version update".format(self._ga_family, self._gs_id)
-                    GAUpdateReportState.report_error_msg = warn_msg
-                agent_manifest = goal_state.fetch_agent_manifest(agent_family.name, agent_family.uris)
-                requested_version = self.__get_largest_version(agent_manifest)
-                self._is_requested_version_update = False
-            else:
-                self._is_requested_version_update = True
-                # Save the requested version to report back
-                GAUpdateReportState.report_expected_version = requested_version
-                # Remove the missing requested version warning once requested version becomes available
-                if "Missing requested version" in GAUpdateReportState.report_error_msg:
-                    GAUpdateReportState.report_error_msg = ""
-
-            # Check if an update is allowed and update next upgrade times even if no new agent is available for upgrade
-            if not self.__check_if_agent_update_allowed_and_update_next_upgrade_times(requested_version):
+            if not conf.get_autoupdate_enabled() or not conf.get_download_new_agents():
                 return
 
-            if requested_version == CURRENT_VERSION:
+            # verify if agent update is allowed this time (RSM checks 1 hr interval; self-update checks manifest download interval)
+            if not self._updater.is_update_allowed_this_time():
                 return
 
-            if warn_msg != "":
-                self.__log_event(LogLevel.WARNING, warn_msg)
-
-            # Downgrades are not allowed for self-update version
-            if not self.__check_if_downgrade_is_requested_and_allowed(requested_version):
+            agent_family = self._get_agent_family_manifest(goal_state)
+            gs_id = goal_state.extensions_goal_state.id
+            try:
+                # updater will raise exception if we need to switch to self-update or rsm update
+                self._updater.check_and_switch_updater_if_changed(agent_family, gs_id)
+            except VMDisabledRSMUpdates:
```

Comment:
Opting in and out of RSM are not error conditions, but the code communicates those transitions with exceptions, which is odd. Exceptions are used to communicate errors. Consider changing that to conditionals.